src/sctools/bam.py

```diff
@@ -213,33 +213,37 @@ def _cleanup(files_to_counts, files_to_names, rm_all=False):
     return list(files_to_names.values())
 
 
-def iter_tag_groups(tag: str, bam_iterator: Iterator[pysam.AlignedSegment]) -> Generator:
+def iter_tag_groups(
+        tag: str,
+        bam_iterator: Iterator[pysam.AlignedSegment],
+        filter_null: bool=False) -> Generator:
+
     # get first read and tag set
     reads = []
-    while not reads:
-
-        try:
-            read = next(bam_iterator)
-            current_tag = read.get_tag(tag)
-            reads.append(read)
-            break
-        except KeyError:  # todo this needs to be tracked so we know how many reads were invalid
-            continue
-        except StopIteration:
-            raise ValueError('No reads contained tag %s' % tag)
+    read = next(bam_iterator)
+    try:
+        current_tag = read.get_tag(tag)
+    except KeyError:
+        current_tag = None  # null tag is a category that gets emitted
 
+    # now iterate over alignment sets
     for alignment in bam_iterator:
         try:
             next_tag = alignment.get_tag(tag)
         except KeyError:
-            continue
+            next_tag = None  # null tag is a category that we will emit
         if next_tag == current_tag:
             reads.append(alignment)
         else:
-            yield iter(reads), current_tag
+            # only yield if the tag is non-null or filter_null is false
+            if not filter_null or current_tag is not None:
```

Comment:
if you do what the comment above suggests you can get rid of the if checks here.

Reply:
I prefer them here because the else clause will execute far less often than the if clause. Thus, faster code.